management/ia/templates_test.go

```diff
@@ -371,60 +371,145 @@ func TestListTemplate(t *testing.T) {
-			assert.Equal(t, float32(0), template.Params[1].Float.Default)
-			assert.False(t, template.Params[1].Float.HasMax)
-			assert.Equal(t, float32(0), template.Params[1].Float.Max)
-			assert.False(t, template.Params[1].Float.HasMin)
-			assert.Equal(t, float32(00), template.Params[1].Float.Min)
-
-			assert.Equal(t, map[string]string{"foo": "bar"}, template.Labels)
-			assert.Equal(t, map[string]string{"description": "test description", "summary": "test summary"}, template.Annotations)
-			assert.Equal(t, yml, template.Yaml)
-			assert.NotEmpty(t, template.CreatedAt)
-			found = true
+	t.Run("with pagination", func(t *testing.T) {
+		const templatesCount = 5
+
+		templateNames := make(map[string]struct{})
+
+		for i := 0; i < templatesCount; i++ {
+			name := gofakeit.UUID()
+			expr := gofakeit.UUID()
+			yml := formatTemplateYaml(t, fmt.Sprintf(string(b), name, expr, "%", "s"))
+			_, err = client.CreateTemplate(&templates.CreateTemplateParams{
+				Body: templates.CreateTemplateBody{
+					Yaml: yml,
+				},
+				Context: pmmapitests.Context,
+			})
+
+			templateNames[name] = struct{}{}
 		}
-	}
-	assert.Truef(t, found, "Template with id %s not found", name)
+		defer func() {
+			for name := range templateNames {
+				deleteTemplate(t, client, name)
+			}
+		}()
+
+		// list rules, so they are all on the first page
+		body := templates.ListTemplatesBody{
+			PageParams: &templates.ListTemplatesParamsBodyPageParams{
+				PageSize: 20,
+				Index:    0,
+			},
+		}
+		list1, err := client.ListTemplates(&templates.ListTemplatesParams{
+			Body:    body,
+			Context: pmmapitests.Context,
+		})
+		require.NoError(t, err)
+
+		assert.GreaterOrEqual(t, len(list1.Payload.Templates), templatesCount)
+		assert.Equal(t, int32(len(list1.Payload.Templates)), list1.Payload.Totals.TotalItems)
+		assert.Equal(t, int32(1), list1.Payload.Totals.TotalPages)
+
+		assertFindTemplate := func(list []*templates.TemplatesItems0, name string) func() bool {
+			return func() bool {
+				for _, tmpl := range list {
+					if tmpl.Name == name {
+						return true
+					}
+				}
+				return false
+			}
+		}
+
+		for name := range templateNames {
+			assert.Conditionf(t, assertFindTemplate(list1.Payload.Templates, name), "template %s not found", name)
+		}
+
+		// paginate page over page with page size 1 and check the order - it should be the same as in list1.
+		// last iteration checks that there is no elements for not existing page.
+		for pageIndex := 0; pageIndex <= len(list1.Payload.Templates); pageIndex++ {
+			body := templates.ListTemplatesBody{
+				PageParams: &templates.ListTemplatesParamsBodyPageParams{
+					PageSize: 1,
+					Index:    int32(pageIndex),
+				},
+			}
+			list2, err := client.ListTemplates(&templates.ListTemplatesParams{
+				Body: body, Context: pmmapitests.Context,
+			})
+			require.NoError(t, err)
+
+			assert.Equal(t, list2.Payload.Totals.TotalItems, list2.Payload.Totals.TotalItems)
```

Comment:
what do we check here?

Reply:
good catch, it should be len of templates

Reply:
actually problem here that we check equality of the same object `list2`